Automated Smoketest Build: failed
Number of Error(s): 11
/mnt/disk1/home/vamosax/smoketest/smoketest-8074/HPCC-Platform/roxie/ccd/ccdserver.cpp:6022:19: error: ‘struct IEngineRowStream’ has no member named ‘start’
/mnt/disk1/home/vamosax/smoketest/smoketest-8074/HPCC-Platform/roxie/ccd/ccdserver.cpp:6052:23: error: ‘struct IEngineRowStream’ has no member named ‘reset’
/mnt/disk1/home/vamosax/smoketest/smoketest-8074/HPCC-Platform/roxie/ccd/ccdserver.cpp:6098:22: error: ‘struct IEngineRowStream’ has no member named ‘querySteppingMeta’
/mnt/disk1/home/vamosax/smoketest/smoketest-8074/HPCC-Platform/roxie/ccd/ccdserver.cpp:15253:34: error: ‘struct IEngineRowStream’ has no member named ‘start’
/mnt/disk1/home/vamosax/smoketest/smoketest-8074/HPCC-Platform/roxie/ccd/ccdserver.cpp:15268:30: error: ‘struct IEngineRowStream’ has no member named ‘reset’
/mnt/disk1/home/vamosax/smoketest/smoketest-8074/HPCC-Platform/roxie/ccd/ccdserver.cpp:15292:36: error: cannot convert ‘IEngineRowStream*’ to ‘IRoxieInput*’ in return
/mnt/disk1/home/vamosax/smoketest/smoketest-8074/HPCC-Platform/roxie/ccd/ccdserver.cpp:6099:5: error: control reaches end of non-void function [-Werror=return-type]
cc1plus: some warnings being treated as errors
make[2]: *** [roxie/ccd/CMakeFiles/ccd.dir/ccdserver.cpp.o] Error 1
make[1]: *** [roxie/ccd/CMakeFiles/ccd.dir/all] Error 2
make: *** [all] Error 2

--- a/common/thorhelper/roxiedebug.hpp
+++ b/common/thorhelper/roxiedebug.hpp
@@ -401,7 +401,7 @@ public:
     virtual IDebugGraphManager *queryDebugManager();
     virtual const char *queryIdString() const;
     virtual unsigned queryId() const;
-    virtual IInputBase *createProbe(IInputBase *in, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
+    virtual IRoxieProbe *createProbe(IInputBase *in, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
     {
         UNIMPLEMENTED;
     }

--- a/common/thorhelper/roxiedebug.ipp
+++ b/common/thorhelper/roxiedebug.ipp
@@ -163,7 +163,7 @@ interface IDebugGraphManager : extends IInterface
 
 interface IProbeManager : public IInterface
 {
-    virtual IInputBase *createProbe(IInputBase *in, IActivityBase *inAct, IActivityBase *outAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration) = 0;
+    virtual IInputBase *createProbe(IInputBase *in, IEngineRowStream *_inStream, IActivityBase *inAct, IActivityBase *outAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration) = 0;
     virtual void getProbeResponse(IPropertyTree *query) = 0;
     virtual void noteSink(IActivityBase *sink) = 0;
     virtual void noteDependency(IActivityBase *sourceActivity, unsigned sourceIndex, unsigned controlId, const char *edgeId, IActivityBase *targetActivity) = 0;

--- a/common/thorhelper/roxiedebug.ipp
+++ b/common/thorhelper/roxiedebug.ipp
@@ -160,10 +160,15 @@ interface IDebugGraphManager : extends IInterface
 };
 
 //==============================================================================================================
+interface IRoxieProbe : public IInterface
+{
+    virtual IInputBase &queryInput() = 0;
+    virtual IEngineRowStream &queryStream() = 0;
+};
 
 interface IProbeManager : public IInterface
 {
-    virtual IInputBase *createProbe(IInputBase *in, IEngineRowStream *_inStream, IActivityBase *inAct, IActivityBase *outAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration) = 0;
+    virtual IRoxieProbe *createProbe(IInputBase *in, IEngineRowStream *_inStream, IActivityBase *inAct, IActivityBase *outAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration) = 0;
     virtual void getProbeResponse(IPropertyTree *query) = 0;
     virtual void noteSink(IActivityBase *sink) = 0;
     virtual void noteDependency(IActivityBase *sourceActivity, unsigned sourceIndex, unsigned controlId, const char *edgeId, IActivityBase *targetActivity) = 0;

--- a/common/thorhelper/roxiehelper.ipp
+++ b/common/thorhelper/roxiehelper.ipp
@@ -33,6 +33,7 @@ struct IInputBase : public IInterface //base for IRoxieInput and IHThorInput
 
     // These will need some thought
     virtual IEngineRowStream &queryStream() = 0;
+
     inline void resetEOF() { queryStream().resetEOF(); }
     inline bool nextGroup(ConstPointerArray & group) { return queryStream().nextGroup(group); }
     inline void readAll(RtlLinkedDatasetBuilder &builder) { return queryStream().readAll(builder); }

--- a/ecl/eclagent/eclagent.cpp
+++ b/ecl/eclagent/eclagent.cpp
@@ -3609,6 +3609,7 @@ class InputProbe : public CInterface, implements IHThorInput, implements IEngine
 {
 protected:
     IHThorInput *in;
+    IEngineRowStream *stream;
     unsigned sourceId;
     unsigned sourceIdx;
     unsigned targetId;
@@ -3625,8 +3626,8 @@ protected:
     bool everStarted;
         
 public:
-    InputProbe(IHThorInput *_in, unsigned _sourceId, unsigned _sourceIdx, unsigned _targetId, unsigned _targetIdx, unsigned _iteration, unsigned _channel)
-        : in(_in), sourceId(_sourceId), sourceIdx(_sourceIdx), targetId(_targetId), targetIdx(_targetIdx), iteration(_iteration), channel(_channel)
+    InputProbe(IHThorInput *_in, IEngineRowStream *_stream, unsigned _sourceId, unsigned _sourceIdx, unsigned _targetId, unsigned _targetIdx, unsigned _iteration, unsigned _channel)
+        : in(_in), stream(_stream), sourceId(_sourceId), sourceIdx(_sourceIdx), targetId(_targetId), targetIdx(_targetIdx), iteration(_iteration), channel(_channel)
     {
         hasStarted = false;
         hasStopped = false;
@@ -3669,12 +3670,12 @@ public:
 
     virtual const void * nextRowGE(const void * seek, unsigned numFields, bool &wasCompleteMatch, const SmartStepExtra &stepExtra)
     {
-        return in->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
+        return stream->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
     }
 
     virtual const void *nextRow()
     {
-        const void *ret = in->nextRow();
+        const void *ret = stream->nextRow();
         if (ret)
         {
             size32_t size = in->queryOutputMeta()->getRecordSize(ret);
@@ -3702,7 +3703,7 @@ public:
     virtual void stop() 
     { 
         hasStopped = true;
-        in->stop(); 
+        stream->stop();
     }
 };
 
@@ -3760,8 +3761,8 @@ class DebugProbe : public InputProbe, implements IActivityDebugContext
     }
 
 public:
-    DebugProbe(IHThorInput *_in, unsigned _sourceId, unsigned _sourceIdx, DebugActivityRecord *_sourceAct, unsigned _targetId, unsigned _targetIdx, DebugActivityRecord *_targetAct, unsigned _iteration, unsigned _channel, IDebuggableContext *_debugContext)
-        : InputProbe(_in, _sourceId, _sourceIdx, _targetId, _targetIdx, _iteration, _channel),
+    DebugProbe(IHThorInput *_in, IEngineRowStream *_stream, unsigned _sourceId, unsigned _sourceIdx, DebugActivityRecord *_sourceAct, unsigned _targetId, unsigned _targetIdx, DebugActivityRecord *_targetAct, unsigned _iteration, unsigned _channel, IDebuggableContext *_debugContext)
+        : InputProbe(_in, _stream, _sourceId, _sourceIdx, _targetId, _targetIdx, _iteration, _channel),
           sourceAct(_sourceAct), targetAct(_targetAct), debugContext(_debugContext)
     {
         historyCapacity = debugContext->getDefaultHistoryCapacity();
@@ -4056,12 +4057,12 @@ public:
 
     virtual const void * nextRowGE(const void * seek, unsigned numFields, bool &wasCompleteMatch, const SmartStepExtra &stepExtra)
     {
-        return in->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
+        return stream->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
     }
 
     virtual void stop()
     {
-        in->stop();
+        stream->stop();
     }
 
     virtual const void *nextRow()
@@ -4161,7 +4162,7 @@ public:
     {
     }
 
-    IInputBase *createProbe(IInputBase *in, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
+    IInputBase *createProbe(IInputBase *in, IEngineRowStream *stream, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
     {
         CriticalBlock b(crit);
         unsigned channel = debugContext->queryChannel();
@@ -4169,7 +4170,7 @@ public:
         unsigned targetId = targetAct->queryId();
         DebugActivityRecord *sourceActRecord = noteActivity(sourceAct, iteration, channel, debugContext->querySequence());
         DebugActivityRecord *targetActRecord = noteActivity(targetAct, iteration, channel, debugContext->querySequence());
-        DebugProbe *probe = new DebugProbe(dynamic_cast<IHThorInput*>(in), sourceId, sourceIdx, sourceActRecord, targetId, targetIdx, targetActRecord, iteration, channel, debugContext);
+        DebugProbe *probe = new DebugProbe(dynamic_cast<IHThorInput*>(in), stream, sourceId, sourceIdx, sourceActRecord, targetId, targetIdx, targetActRecord, iteration, channel, debugContext);
     #ifdef _DEBUG
         DBGLOG("Creating probe for edge id %s in graphManager %p", probe->queryEdgeId(), this);
     #endif

--- a/ecl/eclagent/eclagent.cpp
+++ b/ecl/eclagent/eclagent.cpp
@@ -3605,7 +3605,7 @@ int STARTQUERY_API start_query(int argc, const char *argv[])
 
 //=======================================================================================
 //copied/modified from ccdserver
-class InputProbe : public CInterface, implements IHThorInput, implements IEngineRowStream // base class for the edge probes used for tracing and debugging....
+class InputProbe : public CInterface, implements IHThorInput, implements IEngineRowStream, implements IRoxieProbe // base class for the edge probes used for tracing and debugging....
 {
 protected:
     IHThorInput *in;
@@ -3704,6 +3704,14 @@ public:
     { 
         hasStopped = true;
         stream->stop();
+    }
+    virtual IEngineRowStream &queryStream()
+    {
+        return *this;
+    }
+    virtual IInputBase &queryInput()
+    {
+        return *this;
     }
 };
 
@@ -4162,7 +4170,7 @@ public:
     {
     }
 
-    IInputBase *createProbe(IInputBase *in, IEngineRowStream *stream, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
+    IRoxieProbe *createProbe(IInputBase *in, IEngineRowStream *stream, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
     {
         CriticalBlock b(crit);
         unsigned channel = debugContext->queryChannel();

--- a/ecl/eclagent/eclgraph.cpp
+++ b/ecl/eclagent/eclgraph.cpp
@@ -489,6 +489,7 @@ void EclGraphElement::createActivity(IAgentContext & agent, EclSubGraph * owner)
                     {
                         IInputBase *base = probeManager->createProbe(
                                                         input.queryOutput(branchIndexes.item(i2)),  //input
+                                                        &input.queryOutput(branchIndexes.item(i2))->queryStream(),  //stream
                                                         input.activity.get(),   //Source act
                                                         activity.get(),         //target activity
                                                         0,//input.id, 

--- a/ecl/eclagent/eclgraph.cpp
+++ b/ecl/eclagent/eclgraph.cpp
@@ -487,7 +487,7 @@ void EclGraphElement::createActivity(IAgentContext & agent, EclSubGraph * owner)
                 {
                     if (input.activity)
                     {
-                        IInputBase *base = probeManager->createProbe(
+                        IRoxieProbe *base = probeManager->createProbe(
                                                         input.queryOutput(branchIndexes.item(i2)),  //input
                                                         &input.queryOutput(branchIndexes.item(i2))->queryStream(),  //stream
                                                         input.activity.get(),   //Source act
@@ -495,7 +495,7 @@ void EclGraphElement::createActivity(IAgentContext & agent, EclSubGraph * owner)
                                                         0,//input.id, 
                                                         0,//id, 
                                                         0);
-                        probe = dynamic_cast<IHThorInput *>(base);
+                        probe = & dynamic_cast<IHThorInput &> (base->queryInput());
                     }
                 }
                 else 

--- a/ecl/hthor/hthor.hpp
+++ b/ecl/hthor/hthor.hpp
@@ -57,6 +57,15 @@ struct IHThorInput : public IInputBase
     virtual void updateProgress(IStatisticGatherer &progress) const = 0;
     virtual bool gatherConjunctions(ISteppedConjunctionCollector & collector) { return false; }
     virtual void resetEOF() { }
+
+    // HThor is not going to support parallel streams
+    inline bool nextGroup(ConstPointerArray & group) { return queryStream().nextGroup(group); }
+    inline void readAll(RtlLinkedDatasetBuilder &builder) { return queryStream().readAll(builder); }
+    inline const void *nextRowGE(const void * seek, unsigned numFields, bool &wasCompleteMatch, const SmartStepExtra &stepExtra) { return queryStream().nextRowGE(seek, numFields, wasCompleteMatch, stepExtra); }
+    inline const void *nextRow() { return queryStream().nextRow(); }
+    inline void stop() { queryStream().stop(); }
+    inline const void *ungroupedNextRow() { return queryStream().ungroupedNextRow(); }
+
 };
 
 struct IHThorNWayInput

--- a/roxie/ccd/ccdactivities.cpp
+++ b/roxie/ccd/ccdactivities.cpp
@@ -5164,15 +5164,12 @@ public:
         {
             remoteGraph->beforeExecute();
             Owned<IRoxieInput> input = remoteGraph->startOutput(0, remoteExtractBuilder.size(), remoteExtractBuilder.getbytes(), false);
+            IEngineRowStream &stream = input->queryStream();
             while (!aborted)
             {
-                const void * next = input->nextRow();
+                const void * next = stream.ungroupedNextRow();
                 if (!next)
-                {
-                    next = input->nextRow();
-                    if (!next)
-                        break;
-                }
+                    break;
 
                 size32_t nextSize = meta.getRecordSize(next);
                 //MORE - what about grouping?

--- a/roxie/ccd/ccddebug.cpp
+++ b/roxie/ccd/ccddebug.cpp
@@ -34,6 +34,7 @@ class InputProbe : public CInterface, implements IRoxieInput, implements IEngine
 {
 protected:
     IRoxieInput *in;
+    IEngineRowStream *inStream;
     unsigned sourceId;
     unsigned sourceIdx;
     unsigned targetId;
@@ -51,9 +52,9 @@ protected:
     bool hasStopped;
 
 public:
-    InputProbe(IRoxieInput *_in, IDebuggableContext *_debugContext,
+    InputProbe(IRoxieInput *_in, IEngineRowStream *_inStream, IDebuggableContext *_debugContext,
         unsigned _sourceId, unsigned _sourceIdx, unsigned _targetId, unsigned _targetIdx, unsigned _iteration, unsigned _channel)
-        : in(_in),  debugContext(_debugContext),
+        : in(_in),  inStream(_inStream), debugContext(_debugContext),
           sourceId(_sourceId), sourceIdx(_sourceIdx), targetId(_targetId), targetIdx(_targetIdx), iteration(_iteration), channel(_channel)
     {
         hasStarted = false;
@@ -75,7 +76,7 @@ public:
     }
     virtual void resetEOF()
     {
-        in->resetEOF();
+        inStream->resetEOF();
     }
     virtual unsigned numConcreteOutputs() const
     {
@@ -122,7 +123,7 @@ public:
     virtual void stop()
     {
         hasStopped = true;
-        in->stop();
+        inStream->stop();
     }
     virtual void reset()
     {
@@ -135,7 +136,7 @@ public:
     }
     virtual const void *nextRow()
     {
-        const void *ret = in->nextRow();
+        const void *ret = inStream->nextRow();
         if (ret)
         {
             size32_t size = inMeta->getRecordSize(ret);
@@ -148,7 +149,7 @@ public:
     }
     virtual const void * nextRowGE(const void * seek, unsigned numFields, bool &wasCompleteMatch, const SmartStepExtra & stepExtra)
     {
-        const void *ret = in->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
+        const void *ret = inStream->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
         if (ret && wasCompleteMatch)  // GH is this test right?
         {
             size32_t size = inMeta->getRecordSize(ret);
@@ -167,8 +168,8 @@ class TraceProbe : public InputProbe
 public:
     IMPLEMENT_IINTERFACE;
 
-    TraceProbe(IRoxieInput *_in, unsigned _sourceId, unsigned _targetId, unsigned _sourceIdx, unsigned _targetIdx, unsigned _iteration, unsigned _channel)
-        : InputProbe(_in, NULL, _sourceId, _sourceIdx, _targetId, _targetIdx, _iteration, _channel)
+    TraceProbe(IRoxieInput *_in, IEngineRowStream *_inStream, unsigned _sourceId, unsigned _targetId, unsigned _sourceIdx, unsigned _targetIdx, unsigned _iteration, unsigned _channel)
+        : InputProbe(_in, _inStream, NULL, _sourceId, _sourceIdx, _targetId, _targetIdx, _iteration, _channel)
     {
     }
 
@@ -267,11 +268,11 @@ class CProbeManager : public CInterface, implements IProbeManager
 public:
     IMPLEMENT_IINTERFACE;
 
-    IInputBase *createProbe(IInputBase *in, IActivityBase *inAct, IActivityBase *outAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
+    IInputBase *createProbe(IInputBase *in, IEngineRowStream *_inStream, IActivityBase *inAct, IActivityBase *outAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
     {
         unsigned idIn = inAct->queryId();
         unsigned idOut = outAct->queryId();
-        TraceProbe *probe = new TraceProbe(static_cast<IRoxieInput*>(in), idIn, idOut, sourceIdx, targetIdx, iteration, 0);
+        TraceProbe *probe = new TraceProbe(static_cast<IRoxieInput*>(in), _inStream, idIn, idOut, sourceIdx, targetIdx, iteration, 0);
         probes.append(*probe);
         return probe;
     }
@@ -521,8 +522,8 @@ class DebugProbe : public InputProbe, implements IActivityDebugContext
     }
 
 public:
-    DebugProbe(IInputBase *_in, unsigned _sourceId, unsigned _sourceIdx, DebugActivityRecord *_sourceAct, unsigned _targetId, unsigned _targetIdx, DebugActivityRecord *_targetAct, unsigned _iteration, unsigned _channel, IDebuggableContext *_debugContext)
-        : InputProbe(static_cast<IRoxieInput*>(_in), _debugContext, _sourceId, _sourceIdx, _targetId, _targetIdx, _iteration, _channel),
+    DebugProbe(IInputBase *_in, IEngineRowStream *_inStream, unsigned _sourceId, unsigned _sourceIdx, DebugActivityRecord *_sourceAct, unsigned _targetId, unsigned _targetIdx, DebugActivityRecord *_targetAct, unsigned _iteration, unsigned _channel, IDebuggableContext *_debugContext)
+        : InputProbe(static_cast<IRoxieInput*>(_in), _inStream, _debugContext, _sourceId, _sourceIdx, _targetId, _targetIdx, _iteration, _channel),
           sourceAct(_sourceAct), targetAct(_targetAct)
     {
         historyCapacity = debugContext->getDefaultHistoryCapacity();
@@ -1006,7 +1007,7 @@ public:
         return CInterface::Release();
     }
 
-    virtual IInputBase *createProbe(IInputBase *in, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
+    virtual IInputBase *createProbe(IInputBase *in, IEngineRowStream *inStream, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
     {
         CriticalBlock b(crit);
         if (!iteration)
@@ -1016,7 +1017,7 @@ public:
         unsigned targetId = targetAct->queryId();
         DebugActivityRecord *sourceActRecord = noteActivity(sourceAct, iteration, channel, debugContext->querySequence());
         DebugActivityRecord *targetActRecord = noteActivity(targetAct, iteration, channel, debugContext->querySequence());
-        DebugProbe *probe = new DebugProbe(in, sourceId, sourceIdx, sourceActRecord, targetId, targetIdx, targetActRecord, iteration, channel, debugContext);
+        DebugProbe *probe = new DebugProbe(in, inStream, sourceId, sourceIdx, sourceActRecord, targetId, targetIdx, targetActRecord, iteration, channel, debugContext);
 #ifdef _DEBUG
         DBGLOG("Creating probe for edge id %s in graphManager %p", probe->queryEdgeId(), this);
 #endif

--- a/roxie/ccd/ccddebug.cpp
+++ b/roxie/ccd/ccddebug.cpp
@@ -30,7 +30,7 @@ using roxiemem::IRowManager;
 
 //=======================================================================================================================
 
-class InputProbe : public CInterface, implements IRoxieInput, implements IEngineRowStream // base class for the edge probes used for tracing and debugging....
+class InputProbe : public CInterface, implements IRoxieInput, implements IEngineRowStream, implements IRoxieProbe // base class for the edge probes used for tracing and debugging....
 {
 protected:
     IRoxieInput *in;
@@ -106,6 +106,10 @@ public:
         return in->queryOutputMeta();
     }
     IEngineRowStream &queryStream()
+    {
+        return *this;
+    }
+    IInputBase &queryInput()
     {
         return *this;
     }
@@ -268,7 +272,7 @@ class CProbeManager : public CInterface, implements IProbeManager
 public:
     IMPLEMENT_IINTERFACE;
 
-    IInputBase *createProbe(IInputBase *in, IEngineRowStream *_inStream, IActivityBase *inAct, IActivityBase *outAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
+    IRoxieProbe *createProbe(IInputBase *in, IEngineRowStream *_inStream, IActivityBase *inAct, IActivityBase *outAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
     {
         unsigned idIn = inAct->queryId();
         unsigned idOut = outAct->queryId();
@@ -1007,7 +1011,7 @@ public:
         return CInterface::Release();
     }
 
-    virtual IInputBase *createProbe(IInputBase *in, IEngineRowStream *inStream, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
+    virtual IRoxieProbe *createProbe(IInputBase *in, IEngineRowStream *inStream, IActivityBase *sourceAct, IActivityBase *targetAct, unsigned sourceIdx, unsigned targetIdx, unsigned iteration)
     {
         CriticalBlock b(crit);
         if (!iteration)

--- a/roxie/ccd/ccdserver.cpp
+++ b/roxie/ccd/ccdserver.cpp
@@ -14478,12 +14478,13 @@ private:
     unsigned sourceIdx;
     Linked<IRoxieServerActivity> sourceAct;
     Linked<IRoxieInput> sourceInput;
+    Linked<IEngineRowStream> sourceStream;
     unsigned numUses;
     unsigned iteration;
 
 public:
-    CGraphIterationInfo(IRoxieServerActivity * _sourceAct, IRoxieInput *_input, unsigned _sourceIdx, unsigned _iteration)
-        : sourceAct(_sourceAct), sourceInput(_input), sourceIdx(_sourceIdx), iteration(_iteration)
+    CGraphIterationInfo(IRoxieServerActivity * _sourceAct, IRoxieInput *_input, IEngineRowStream *_stream, unsigned _sourceIdx, unsigned _iteration)
+        : sourceAct(_sourceAct), sourceInput(_input), sourceStream(_stream),  sourceIdx(_sourceIdx), iteration(_iteration)
     {
         numUses = 0;
     }
@@ -14503,7 +14504,7 @@ public:
             IRoxieInput *input = sourceAct->queryOutput(sourceIdx);
             if (probeManager)
             {
-                IInputBase * inputBase = probeManager->createProbe(static_cast<IInputBase*>(input), sourceAct, splitter, sourceIdx, 0, iteration);
+                IInputBase * inputBase = probeManager->createProbe(static_cast<IInputBase*>(input), sourceStream, sourceAct, splitter, sourceIdx, 0, iteration);
                 input = static_cast<IRoxieInput*>(inputBase);
                 // MORE - shouldn't this be added to probes?
             }
@@ -14511,6 +14512,7 @@ public:
             sourceAct->setInput(0, input);
             sourceIdx = 0;
             sourceInput.clear();
+            sourceStream.clear();
         }
     }
 
@@ -14518,11 +14520,14 @@ public:
     {
         // MORE - not really necessary to create splitters in separate pass, is it?
         if (factory) // we created a splitter....
+        {
             sourceInput.set(sourceAct->queryOutput(sourceIdx));
+            sourceStream.set(&sourceInput->queryStream());
+        }
         IRoxieInput *ret = sourceInput;
         if (probeManager)
         {
-            IInputBase *inputBase = probeManager->createProbe(ret, sourceAct, targetAct, sourceIdx, targetIdx, iteration);
+            IInputBase *inputBase = probeManager->createProbe(sourceInput, sourceStream, sourceAct, targetAct, sourceIdx, targetIdx, iteration);
             ret = static_cast<IRoxieInput *>(inputBase);
             probes.append(*LINK(ret));
         }
@@ -14625,7 +14630,7 @@ public:
     {
         //result(0) is the input to the graph.
         resultInput = inputExtractMapper;
-        outputs.append(* new CGraphIterationInfo(resultInput->queryActivity(), resultInput, 0, 1));
+        outputs.append(* new CGraphIterationInfo(resultInput->queryActivity(), resultInput, &resultInput->queryStream(), 0, 1));
 
         for (createLoopCounter=1; createLoopCounter <= maxIterations; createLoopCounter++)
         {
@@ -25695,7 +25700,7 @@ public:
         IRoxieInput * output = sourceActivity.queryOutput(sourceIdx);
         if (probeManager)
         {
-            IInputBase * inputBase = probeManager->createProbe(static_cast<IInputBase*>(output), &sourceActivity, &targetActivity, sourceIdx, targetIdx, iteration);
+            IInputBase * inputBase = probeManager->createProbe(static_cast<IInputBase*>(output), &output->queryStream(), &sourceActivity, &targetActivity, sourceIdx, targetIdx, iteration);
             output = static_cast<IRoxieInput*>(inputBase);
             probes.append(*LINK(output));
         }
@@ -26136,7 +26141,9 @@ public:
     virtual CGraphIterationInfo *selectGraphLoopOutput()
     {
         IRoxieServerActivity &sourceActivity = activities.item(graphOutputActivityIndex);
-        return new CGraphIterationInfo(&sourceActivity, sourceActivity.queryOutput(0), 0, loopCounter);
+        IRoxieInput *sourceInput = sourceActivity.queryOutput(0);
+        IEngineRowStream *sourceStream = &sourceInput->queryStream();
+        return new CGraphIterationInfo(&sourceActivity, sourceInput, sourceStream, 0, loopCounter);
     }
 
     virtual void gatherIterationUsage(IRoxieServerLoopResultProcessor & processor)

--- a/roxie/ccd/ccdserver.cpp
+++ b/roxie/ccd/ccdserver.cpp
@@ -1350,7 +1350,7 @@ public:
     {
     }
 
-    virtual void associateIterationOutputs(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract, IProbeManager *probeManager, IArrayOf<IRoxieInput> &probes)
+    virtual void associateIterationOutputs(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract, IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes)
     {
     }
 
@@ -5351,9 +5351,9 @@ interface ILocalGraphEx : public IEclGraphResults
 {
 public:
     virtual void setResult(unsigned id, IGraphResult * result) = 0;
-    virtual IRoxieInput * createResultIterator(unsigned id) = 0;
+    virtual IEngineRowStream * createResultIterator(unsigned id) = 0;
     virtual void setGraphLoopResult(IGraphResult * result) = 0;
-    virtual IRoxieInput * createGraphLoopResultIterator(unsigned id) = 0;
+    virtual IEngineRowStream * createGraphLoopResultIterator(unsigned id) = 0;
 };
 
 
@@ -5595,7 +5595,7 @@ public:
     }
 
 // interface IGraphResult
-    virtual IRoxieInput * createIterator()
+    virtual IEngineRowStream * createIterator()
     {
         if (!complete)
             throw MakeStringException(ROXIE_GRAPH_PROCESSING_ERROR, "Internal Error: Reading uninitialised graph result"); 
@@ -5635,21 +5635,27 @@ public:
     }
 
 protected:
-    class CGraphResultIterator : public CPseudoRoxieInput
+    class CGraphResultIterator : public CInterfaceOf<IEngineRowStream>
     {
         unsigned i;
         Linked<CGraphResult> result;
 
     public:
         CGraphResultIterator(CGraphResult * _result) : result(_result) { i = 0; }
-        IMPLEMENT_IINTERFACE
-
 
     public:
         virtual const void * nextRow()
         {
             return result->getRow(i++);
         }
+        virtual void stop()
+        {
+        }
+        virtual void resetEOF()
+        {
+            throwUnexpected();
+        }
+
     };
 };
 
@@ -5660,7 +5666,7 @@ class CRoxieServerLocalResultReadActivity : public CRoxieServerActivity
 {
     IHThorLocalResultReadArg &helper;
     CriticalSection iterCrit;
-    Owned<IRoxieInput> iter;
+    Owned<IEngineRowStream> iter;
     ILocalGraphEx * graph;
     unsigned graphId;
     unsigned sequence;
@@ -5698,7 +5704,7 @@ public:
     virtual const void *nextRow()
     {
         ActivityTimer t(totalCycles, timeActivities);
-        Linked<IRoxieInput> useIter;
+        Linked<IEngineRowStream> useIter;
         {
             CriticalBlock b(iterCrit);
             if (!iter)
@@ -5990,7 +5996,8 @@ class CRoxieServerGraphLoopResultReadActivity : public CRoxieServerActivity
 protected:
     IHThorGraphLoopResultReadArg &helper;
     CriticalSection iterCrit;
-    Owned<IRoxieInput> iter;
+    Owned<IRoxieInput> iterInput;
+    Owned<IEngineRowStream> iterStream;
     ILocalGraphEx * graph;
     unsigned graphId;
     unsigned sequence;
@@ -6012,8 +6019,8 @@ public:
     virtual void start(unsigned parentExtractSize, const byte *parentExtract, bool paused)
     {
         CRoxieServerActivity::start(parentExtractSize, parentExtract, paused);
-        if (iter)
-            iter->start(parentExtractSize, parentExtract, paused);
+        if (iterInput)
+            iterInput->start(parentExtractSize, parentExtract, paused);
         else
         {
             sequence = helper.querySequence();
@@ -6021,7 +6028,8 @@ public:
             {
                 try
                 {
-                    iter.setown(graph->createGraphLoopResultIterator(sequence));
+                    iterStream.setown(graph->createGraphLoopResultIterator(sequence));
+                    iterInput.clear();
                 }
                 catch (IException * E)
                 {
@@ -6033,8 +6041,8 @@ public:
 
     virtual void stop()
     {
-        if (iter)
-            iter->stop();
+        if (iterStream)
+            iterStream->stop();
         CRoxieServerActivity::stop();
     }
 
@@ -6042,9 +6050,10 @@ public:
     {
         {
             CriticalBlock b(iterCrit);
-            if (iter)
-                iter->reset();
-            iter.clear();
+            if (iterInput)
+                iterInput->reset();
+            iterInput.clear();
+            iterStream.clear();
         }
         CRoxieServerActivity::reset(); 
     };
@@ -6052,14 +6061,14 @@ public:
     virtual const void *nextRow()
     {
         ActivityTimer t(totalCycles, timeActivities);
-        Linked<IRoxieInput> useIter;
+        Linked<IEngineRowStream> useStream;
         {
             CriticalBlock b(iterCrit);
-            if (!iter)
+            if (!iterStream)
                 return NULL;
-            useIter.set(iter);
+            useStream.set(iterStream);
         }
-        const void * next = useIter->nextRow();
+        const void * next = useStream->nextRow();
         if (next)
         {
             processed++;
@@ -6080,22 +6089,26 @@ public:
         processor.noteUseIteration(helper.querySequence());
     }
 
-    virtual void associateIterationOutputs(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract, IProbeManager *probeManager, IArrayOf<IRoxieInput> &probes)
+    virtual void associateIterationOutputs(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract, IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes)
     {
-        //helper already initialised from the gratherIterationUsage() call.
-        iter.set(processor.connectIterationOutput(helper.querySequence(), probeManager, probes, this, 0));
+        //helper already initialised from the gatherIterationUsage() call.
+        iterInput.set(processor.connectIterationOutput(helper.querySequence(), probeManager, probes, this, 0));
+        if (iterInput)
+            iterStream.set(&iterInput->queryStream());
+        else
+            iterStream.clear();
     }
 
     virtual IInputSteppingMeta * querySteppingMeta()
     {
-        assertex(iter);
-        return iter->querySteppingMeta();
+        assertex(iterInput);
+        return iterInput->querySteppingMeta();
     }
 
     virtual const void * nextRowGE(const void * seek, unsigned numFields, bool &wasCompleteMatch, const SmartStepExtra & stepExtra)
     {
-        assertex(iter);
-        return iter->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
+        assertex(iterStream);
+        return iterStream->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
     }
 };
 
@@ -6117,7 +6130,8 @@ public:
         {
             try
             {
-                iter.setown(graph->createGraphLoopResultIterator(sequence));
+                iterStream.setown(graph->createGraphLoopResultIterator(sequence));
+                iterInput.clear();
             }
             catch (IException * E)
             {
@@ -13623,17 +13637,17 @@ class CRoxieServerSequentialLoopActivity : public CRoxieServerLoopActivity
 {
     Owned<IActivityGraph> loopQuery;
     Owned<IRoxieServerChildGraph> loopGraph;
-    IRoxieInput * curInput;
+    IEngineRowStream * curStream;
     RtlLinkedDatasetBuilder *loopInputBuilder;
     CPointerArrayRoxieInput arrayInput;
-    Linked<IRoxieInput> resultInput; 
+    Linked<IEngineRowStream> resultStream;
     unsigned loopCounter;
 
 public:
     CRoxieServerSequentialLoopActivity(const IRoxieServerActivityFactory *_factory, IProbeManager *_probeManager, unsigned _loopGraphId, IOutputMetaData * _counterMeta)
         : CRoxieServerLoopActivity(_factory, _probeManager, _loopGraphId, _counterMeta)
     {
-        curInput = NULL;
+        curStream = NULL;
         loopCounter = 0;
         loopInputBuilder = NULL;
     }
@@ -13648,7 +13662,7 @@ public:
 
     virtual void start(unsigned parentExtractSize, const byte *parentExtract, bool paused)
     {
-        curInput = input;
+        curStream = &input->queryStream();
         loopCounter = 1;
         CRoxieServerLoopActivity::start(parentExtractSize, parentExtract, paused);
 
@@ -13674,10 +13688,10 @@ public:
         {
             loop
             {
-                const void * ret = curInput->nextRow();
+                const void * ret = curStream->nextRow();
                 if (!ret)
                 {
-                    ret = curInput->nextRow();      // more cope with groups somehow....
+                    ret = curStream->nextRow();      // more cope with groups somehow....
                     if (!ret)
                     {
                         if (finishedLooping)
@@ -13713,7 +13727,7 @@ public:
                             }
                             arrayInput.init(loopInputBuilder->getcount(), loopInputBuilder->linkrows());
                             // MORE - should builder be cleared here?
-                            curInput = &arrayInput;
+                            curStream = &arrayInput;
                             finishedLooping = true;
                             continue;       // back to the input loop again
                         }
@@ -13746,11 +13760,11 @@ public:
             try 
             {
                 Owned<IRoxieGraphResults> results = executeIteration(loopExtractBuilder.size(), loopExtractBuilder.getbytes(), loopCounter);
-                resultInput.setown(results->createIterator(0));
+                resultStream.setown(results->createIterator(0));
 
                 if (flags & IHThorLoopArg::LFnewloopagain)
                 {
-                    Owned<IRoxieInput> againResult = results->createIterator(helper.loopAgainResult());
+                    Owned<IEngineRowStream> againResult = results->createIterator(helper.loopAgainResult());
                     OwnedConstRoxieRow row  = againResult->nextRow();
                     assertex(row);
                     //Result is a row which contains a single boolean field.
@@ -13763,7 +13777,7 @@ public:
                 throw makeWrappedException(E);
             }
 
-            curInput = resultInput.get();
+            curStream = resultStream.get();
 
             loopCounter++;
             if ((activityKind == TAKloopcount) && (loopCounter > maxIterations))
@@ -14367,7 +14381,7 @@ class CRoxieServerSequentialGraphLoopActivity : public CRoxieServerGraphLoopActi
 {
     Owned<IActivityGraph> GraphQuery;
     Owned<IRoxieServerChildGraph> loopGraph;
-    Linked<IRoxieInput> resultInput; 
+    Linked<IEngineRowStream> resultStream;
     bool evaluated;
 
 public:
@@ -14408,7 +14422,7 @@ public:
             evaluated = true;
         }
 
-        const void * ret = resultInput->nextRow();
+        const void * ret = resultStream->nextRow();
         if (ret)
             processed++;
         return ret;
@@ -14450,7 +14464,7 @@ public:
             executeIteration(GraphExtractBuilder.size(), GraphExtractBuilder.getbytes(), loopCounter);
         }
 
-        resultInput.setown(loopGraph->getGraphLoopResult(maxIterations));
+        resultStream.setown(loopGraph->getGraphLoopResult(maxIterations));
     }
 };
 
@@ -14494,7 +14508,7 @@ public:
         numUses++;
     }
 
-    void createSplitter(IRoxieSlaveContext *ctx, IProbeManager *probeManager)
+    void createSplitter(IRoxieSlaveContext *ctx, IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes)
     {
         if (numUses > 1)
         {
@@ -14504,9 +14518,9 @@ public:
             IRoxieInput *input = sourceAct->queryOutput(sourceIdx);
             if (probeManager)
             {
-                IInputBase * inputBase = probeManager->createProbe(static_cast<IInputBase*>(input), sourceStream, sourceAct, splitter, sourceIdx, 0, iteration);
-                input = static_cast<IRoxieInput*>(inputBase);
-                // MORE - shouldn't this be added to probes?
+                IRoxieProbe * inputProbe = probeManager->createProbe(static_cast<IInputBase*>(input), sourceStream, sourceAct, splitter, sourceIdx, 0, iteration);
+                probes.append(*LINK(inputProbe));
+                input = &dynamic_cast<IRoxieInput &>(inputProbe->queryInput());
             }
             sourceAct.setown(splitter);
             sourceAct->setInput(0, input);
@@ -14516,7 +14530,7 @@ public:
         }
     }
 
-    IRoxieInput *connectOutput(IProbeManager *probeManager, IArrayOf<IRoxieInput> &probes, IRoxieServerActivity *targetAct, unsigned targetIdx)
+    IRoxieInput *connectOutput(IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes, IRoxieServerActivity *targetAct, unsigned targetIdx)
     {
         // MORE - not really necessary to create splitters in separate pass, is it?
         if (factory) // we created a splitter....
@@ -14527,9 +14541,9 @@ public:
         IRoxieInput *ret = sourceInput;
         if (probeManager)
         {
-            IInputBase *inputBase = probeManager->createProbe(sourceInput, sourceStream, sourceAct, targetAct, sourceIdx, targetIdx, iteration);
-            ret = static_cast<IRoxieInput *>(inputBase);
-            probes.append(*LINK(ret));
+            IRoxieProbe * inputProbe = probeManager->createProbe(sourceInput, sourceStream, sourceAct, targetAct, sourceIdx, targetIdx, iteration);
+            probes.append(*LINK(inputProbe));
+            ret = &dynamic_cast<IRoxieInput &>(inputProbe->queryInput());
         }
         if (factory) // we created a splitter....
             sourceIdx++;
@@ -14542,14 +14556,14 @@ public:
 class CRoxieServerParallelGraphLoopActivity : public CRoxieServerGraphLoopActivity, implements IRoxieServerLoopResultProcessor
 {
     Owned<IActivityGraph> childGraph;
-    IRoxieInput * resultInput; 
+    IRoxieInput * resultInput;
     CIArrayOf<CGraphIterationInfo> outputs;
     IArrayOf<IRoxieServerChildGraph> iterationGraphs;
     Owned<CExtractMapperInput> inputExtractMapper;
     IProbeManager *probeManager;
     unsigned createLoopCounter;
 
-    IArrayOf<IRoxieInput> probes;
+    IArrayOf<IRoxieProbe> probes;
 
 public:
     CRoxieServerParallelGraphLoopActivity(const IRoxieServerActivityFactory *_factory, IProbeManager *_probeManager, unsigned _GraphGraphId, IOutputMetaData * _counterMeta)
@@ -14643,19 +14657,19 @@ public:
         }
         createLoopCounter = 0;
 
-        createSplitters(probeManager);
+        createSplitters(probeManager, probes);
 
         ForEachItemIn(i2, iterationGraphs)
             iterationGraphs.item(i2).associateIterationOutputs(*this);
         resultInput = outputs.tos().connectOutput(probeManager, probes, this, 0);
     }
 
-    void createSplitters(IProbeManager *probeManager)
+    void createSplitters(IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes)
     {
         ForEachItemIn(i, outputs)
         {
             CGraphIterationInfo & next = outputs.item(i);
-            next.createSplitter(ctx, probeManager);
+            next.createSplitter(ctx, probeManager, probes);
         }
     }
 
@@ -14671,7 +14685,7 @@ public:
         }
     }
 
-    virtual IRoxieInput * connectIterationOutput(unsigned whichIteration, IProbeManager *probeManager, IArrayOf<IRoxieInput> &probes, IRoxieServerActivity *targetAct, unsigned targetIdx)
+    virtual IRoxieInput* connectIterationOutput(unsigned whichIteration, IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes, IRoxieServerActivity *targetAct, unsigned targetIdx)
     {
         if (outputs.isItem(whichIteration))
         {
@@ -15299,9 +15313,9 @@ public:
             processor.noteUseIteration(selections[i]);
     }
 
-    virtual void associateIterationOutputs(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract, IProbeManager *probeManager, IArrayOf<IRoxieInput> &probes)
+    virtual void associateIterationOutputs(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract, IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes)
     {
-        //selection  etc. already initialised from the gratherIterationUsage() call.
+        //selection  etc. already initialised from the gatherIterationUsage() call.
         unsigned max = selectionLen / sizeof(size32_t);
         const size32_t * selections = (const size32_t *)selection.getdata();
         for (unsigned i = 0; i < max; i++)
@@ -25441,7 +25455,7 @@ public:
         results.kill();
     }
 
-    IRoxieInput * createIterator(unsigned id)
+    virtual IEngineRowStream * createIterator(unsigned id)
     {
         return select(id).createIterator();
     }
@@ -25611,7 +25625,7 @@ protected:
     // NOTE - destructor order is significant - need to destroy graphCodeContext and graphSlaveContext last
 
     IArrayOf<IRoxieServerActivity> activities;
-    IArrayOf<IRoxieInput> probes;
+    IArrayOf<IRoxieProbe> probes;
     IRoxieServerActivityCopyArray sinks;
     StringAttr graphName;
     Owned<CGraphResults> results;
@@ -25700,9 +25714,9 @@ public:
         IRoxieInput * output = sourceActivity.queryOutput(sourceIdx);
         if (probeManager)
         {
-            IInputBase * inputBase = probeManager->createProbe(static_cast<IInputBase*>(output), &output->queryStream(), &sourceActivity, &targetActivity, sourceIdx, targetIdx, iteration);
-            output = static_cast<IRoxieInput*>(inputBase);
-            probes.append(*LINK(output));
+            IRoxieProbe * inputProbe = probeManager->createProbe(static_cast<IInputBase*>(output), &output->queryStream(), &sourceActivity, &targetActivity, sourceIdx, targetIdx, iteration);
+            probes.append(*LINK(inputProbe));
+            output = &dynamic_cast<IRoxieInput &>(inputProbe->queryInput());
         }
         targetActivity.setInput(targetIdx, output);
     }
@@ -25938,7 +25952,7 @@ public:
     {
         results->setResult(id, result);
     }
-    virtual IRoxieInput * createResultIterator(unsigned id)
+    virtual IEngineRowStream * createResultIterator(unsigned id)
     {
         return results->createIterator(id);
     }
@@ -25946,7 +25960,7 @@ public:
     {
         graphLoopResults.appendResult(result);
     }
-    virtual IRoxieInput * createGraphLoopResultIterator(unsigned id)
+    virtual IEngineRowStream * createGraphLoopResultIterator(unsigned id)
     {
         try
         {
@@ -25970,7 +25984,7 @@ public:
     {
         graphLoopResults.setResult(id, result);
     }
-    virtual IRoxieInput * getGraphLoopResult(unsigned id)
+    virtual IEngineRowStream * getGraphLoopResult(unsigned id)
     {
         return graphLoopResults.createIterator(id);
     }

--- a/roxie/ccd/ccdserver.hpp
+++ b/roxie/ccd/ccdserver.hpp
@@ -156,7 +156,7 @@ interface IRoxieServerActivity : extends IActivityBase
     virtual IRoxieInput * querySelectOutput(unsigned id) = 0;
     virtual bool querySetStreamInput(unsigned id, IRoxieInput * _input) = 0;
     virtual void gatherIterationUsage(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract) = 0;
-    virtual void associateIterationOutputs(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract, IProbeManager *probeManager, IArrayOf<IRoxieInput> &probes) = 0;
+    virtual void associateIterationOutputs(IRoxieServerLoopResultProcessor & processor, unsigned parentExtractSize, const byte * parentExtract, IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes) = 0;
     virtual void resetOutputsUsed() = 0;        // use for adjusting correct number of uses for a splitter
     virtual void noteOutputUsed() = 0;
     virtual bool isPassThrough() = 0;
@@ -209,20 +209,20 @@ interface IRoxieServerActivityFactory : extends IActivityFactory
 interface IGraphResult : public IInterface
 {
     virtual void getLinkedResult(unsigned & countResult, byte * * & result) = 0;
-    virtual IRoxieInput * createIterator() = 0;
+    virtual IEngineRowStream * createIterator() = 0;
     virtual const void * getLinkedRowResult() = 0;
 };
 
 interface IRoxieServerLoopResultProcessor
 {
     virtual void noteUseIteration(unsigned whichIteration) = 0;
-    virtual IRoxieInput * connectIterationOutput(unsigned whichIteration, IProbeManager *probeManager, IArrayOf<IRoxieInput> &probes, IRoxieServerActivity *targetAct, unsigned targetIdx) = 0;
+    virtual IRoxieInput * connectIterationOutput(unsigned whichIteration, IProbeManager *probeManager, IArrayOf<IRoxieProbe> &probes, IRoxieServerActivity *targetAct, unsigned targetIdx) = 0;
 };
 
 interface IRoxieGraphResults : extends IEclGraphResults
 {
 public:
-    virtual IRoxieInput * createIterator(unsigned id) = 0;
+    virtual IEngineRowStream * createIterator(unsigned id) = 0;
 };
 
 class CGraphIterationInfo;
@@ -241,7 +241,7 @@ interface IRoxieServerChildGraph : public IInterface
     virtual void clearGraphLoopResults() = 0;
     virtual void executeGraphLoop(size32_t parentExtractSize, const byte *parentExtract) = 0;
     virtual void setGraphLoopResult(unsigned id, IGraphResult * result) = 0;
-    virtual IRoxieInput * getGraphLoopResult(unsigned id) = 0;
+    virtual IEngineRowStream * getGraphLoopResult(unsigned id) = 0;
 //parallel graph related helpers.
     virtual CGraphIterationInfo * selectGraphLoopOutput() = 0;
     virtual void gatherIterationUsage(IRoxieServerLoopResultProcessor & processor) = 0;

